src/Elasticsearch/Metadata/Resource/Factory/ElasticsearchProviderResourceMetadataCollectionFactory.php

```diff
@@ -80,21 +62,70 @@ public function create(string $resourceClass): ResourceMetadataCollection
         return $resourceMetadataCollection;
     }
 
-    private function hasIndices(Operation $operation): bool
+    private static function guessIndexName(Operation $operation): string
     {
-        if (false === $operation->getElasticsearch()) {
-            return false;
+        return Inflector::tableize($operation->getShortName());
+    }
+
+    private function configureOperation(Operation $operation, string $resourceClass): ?ElasticsearchOperation
+    {
+        if (!$operation instanceof HttpOperation) {
+            return null;
         }
 
-        $shortName = $operation->getShortName();
-        $index = Inflector::tableize($shortName);
+        $isCollection = $operation instanceof CollectionOperationInterface;
 
-        try {
-            $this->client->cat()->indices(['index' => $index]);
+        if ($operation instanceof ElasticsearchOperation) {
+            if (false === $operation->getElasticsearch()) {
+                throw new \LogicException(sprintf('You cannot disable elasticsearch with %s, use %s instead', ElasticsearchOperation::class, Operation::class));
+            }
+            if (null === $operation->getIndex()) {
+                $operation = $operation->withIndex(self::guessIndexName($operation));
+            }
+        } else {
+            if (false === $operation->getElasticsearch() || null !== $operation->getProvider()) {
+                return null;
+            }
+            if (null === $operation->getElasticsearch() && !$this->canFindIndex(self::guessIndexName($operation))) {
+                return null;
+            }
 
-            return true;
-        } catch (Missing404Exception) {
-            return false;
+            // 1. mapping
+            $indexName = $this->mapping[$resourceClass]['index'] ?? null;
+            $type = $this->mapping[$resourceClass]['type'] ?? null;
+
+            // 2. attribute
+            $extraProperties = $operation->getExtraProperties();
+            $indexName ??= $extraProperties['elasticsearch_index'] ?? null;
+            $type ??= $extraProperties['elasticsearch_type'] ?? null;
+            if ((isset($extraProperties['elasticsearch_index']) || isset($extraProperties['elasticsearch_type'])) && null === $operation->getElasticsearch()) {
+                trigger_deprecation('api-platform/core', '3.1', 'The extra properties "elasticsearch_index" and "elasticsearch_type" are deprecated. Configure %s or %s instead.', Get::class, GetCollection::class);
+            }
```

Comment:
I don't get `extraProperties` here, we should have a `LegacyElasticsearchProviderResourceMetadataCollectionFactory` to handle using `Get` with an `elasticsearch` argument to split the logic between the two.